File: 1300.sum_of_mutated_array_closest_to_target/solution.go
package leetcode1300

import "github.com/Ackerr/Algorithms/utils"

func findBestValue(arr []int, target int) int {
	left := 0
	right := utils.MaxList(arr)
	for left <= right {
		mid := left + ((right - left) >> 1)
		cur := targetSum(arr, mid)
		if cur > target {
			right = mid - 1
		} else if cur < target {
			left = mid + 1
		} else {
			return mid
		}
	}
	ansLeft := targetSum(arr, left)
	ansRight := targetSum(arr, right)
	if utils.Abs(ansLeft-target) >= utils.Abs(ansRight-target) {
		return right
	}
	return left
}

func targetSum(array []int, target int) int {
	sum := 0
	for _, v := range array {
		if v > target {
			sum += target
		} else {
			sum += v
		}
	}
	return sum
}

File: 1300.sum_of_mutated_array_closest_to_target/solution.py
from typing import List


class Solution:
    def findBestValue(self, arr: List[int], target: int) -> int:
        left = 0
        right = max(arr)
        while left <= right:
            mid = left + ((right - left) >> 1)
            cur = sum(mid if i > mid else i for i in arr)
            if cur < target:
                left = mid + 1
            elif cur > target:
                right = mid - 1
            else:
                return mid
        ans1 = sum(left if i > left else i for i in arr)
        ans2 = sum(right if i > right else i for i in arr)
        return right if abs(ans1 - target) >= abs(ans2 - target) else left
